src/librustc/infer/error_reporting/anon_anon_conflict.rs

```diff
@@ -0,0 +1,197 @@
+                    (self.find_arg_with_anonymous_region(sup, sup),
+                     self.find_arg_with_anonymous_region(sub, sub)) {
+
+                    let span_label_var1 = if let Some(simple_name) = anon_arg1.pat.simple_name() {
+                        format!(" from `{}` ", simple_name)
+                    } else {
+                        format!(" ")
+                    };
+
+                    let span_label_var2 = if let Some(simple_name) = anon_arg2.pat.simple_name() {
+                        format!(" into `{}` ", simple_name)
+                    } else {
+                        format!(" ")
+                    };
+
+                    struct_span_err!(self.tcx.sess, span, E0623, "lifetime mismatch")
+                        .span_label(ty1.span,
+                                    format!("these references must have the same lifetime"))
+                        .span_label(ty2.span, format!(""))
+                        .span_label(span,
+                                    format!("data{}flows{}here", span_label_var1, span_label_var2))
+                        .emit();
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
+    pub fn is_anonymous_region(&self, region: Region<'tcx>) -> Option<ty::BoundRegion> {
+
+        match *region {
+            ty::ReFree(ref free_region) => {
+                match free_region.bound_region {
+                    ty::BrAnon(..) => {
+                        let anonymous_region_binding_scope = free_region.scope;
+                        let node_id = self.tcx
+                            .hir
+                            .as_local_node_id(anonymous_region_binding_scope)
+                            .unwrap();
+                        match self.tcx.hir.find(node_id) {
+                            Some(hir_map::NodeItem(..)) |
+                            Some(hir_map::NodeTraitItem(..)) => {
+                                // proceed ahead //
+                            }
+                            Some(hir_map::NodeImplItem(..)) => {
+                                let container_id = self.tcx
+                                    .associated_item(anonymous_region_binding_scope)
+                                    .container
+                                    .id();
+                                if self.tcx.impl_trait_ref(container_id).is_some() {
+                                    // For now, we do not try to target impls of traits. This is
+                                    // because this message is going to suggest that the user
+                                    // change the fn signature, but they may not be free to do so,
+                                    // since the signature must match the trait.
+                                    //
+                                    // FIXME(#42706) -- in some cases, we could do better here.
+                                    return None;
```

Comment:
In general, this function feels familiar -- I think we discussed this already, can it be combined with code from your other PR?